# Variant calling Analysis

Variant calling analysis runs the following workflows:

- Mutationseq
- Strelka
- Mutect

The analysis takes in one or more tumour files and a single normal bam file as input.  The above tools are used to predict somatic SNVs and InDels, with the output being a set of variant calls in VCF per input tumour BAM.  Multiple tumour bam inputs are provided as a way of allowing for the input of sequencing from multiple scWGS runs, or as a way to split very large datasets into managable sets of bams.  VCFs from each tumour BAM and each variant calling tool are input into a consensus algorithm to identify a set of high confidence variant calls.  Consensus VCFs from each tumour BAM are merged (union set) and converted to MAF using vcf2maf.  The workflow outputs a vcf per tumour bam input, per analysis type, a consensus VCF per tumour BAM, and a single MAF file of high confidence SNVs and InDels.

![](https://lucid.app/publicSegments/view/863a76d0-b7ef-408c-bb08-07597fd9fda8/image.png)

## Input Data Format:

### Tumour:

A BAM file:
- generated by the `merge_cells` analysis from mondrian, as explained [here](data_formats/merged_library_bam.md) or
- single bam file with all tumour cells merged, and one read group per sample

### Normal:

A normal BAM file:
- generated by the `merge_cells` analysis on normal cell bams from mondrian as explained [here](data_formats/merged_library_bam.md), or
- single bam file with all normal cells merged, and one read group per sample, or
- bam file generated by bulk WGS sequencing

### Output Data Format:

- {sample}/museq.vcf
- {sample}/mutect.vcf
- {sample}/strelka_snv.csv
- {sample}/strelka_indel.vcf
- {sample}/somatic.vcf
- high_confidence.csv.gz
- somatic.maf
