# Variant calling Analysis

Variant calling analysis runs the following workflows:

- Mutationseq
- Strelka
- Mutect

The analysis takes in one or more tumour files and a single normal bam file as input.  The above tools are used to predict somatic SNVs and InDels, with the output being a set of variant calls.  Multiple tumour bam inputs are provided as a way of allowing for the input of sequencing from multiple scWGS runs, or as a way to split very large datasets into managable sets of bams.  The workflow outputs a vcf per tumour bam input, per analysis type.  In addition, vcfs are merged and converted to maf to produce a single somatic high confidence maf of SNVs and InDels.

> Note that currently only one bam input is supported.

![](https://lucid.app/publicSegments/view/d681122f-8d6c-4eec-8316-2dab52aa4396/image.png)

## Input Data Format:

### Tumour:

A Bam file:
- generated by the `merge_cells` analysis from mondrian, as explained [here](data_formats/merged_library_bam.md) or
- single bam file with all tumour cells merged, and one read group per sample


### Normal:

A normal Bam file:
- generated by the `merge_cells` analysis on normal cell bams from mondrian as explained [here](data_formats/merged_library_bam.md), or
- single bam file with all normal cells merged, and one read group per sample, or
- bam file generated by bulk WGS sequencing


### Output Data Format:

- museq.vcf
- mutect.vcf
- strelka_snv.csv
- strelka_indel.vcf
- high_confidence.csv.gz
- somatic.maf (future)
